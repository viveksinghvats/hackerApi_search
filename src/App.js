
import React from 'react';
import axios from 'axios';
import { result, sortBy } from 'lodash';
import styles from './App.module.css';

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  React.useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);

  return [value, setValue];
};

const storiesReducer = (state, action) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          story => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

// const getLastSearches=urls=>urls.slice(-5).map(extractLastSearchTerm);

const extractLastSearchTerm=url=>url.replace(API_ENDPOINT,'');
// const getLastSearches=urls=>urls.slice(-5).map(url=>extractLastSearchTerm(url));
const getUrl=searchTerm=>`${API_ENDPOINT}${searchTerm}`;

const getLastSearches=urls=>urls.reduce((result,url,index)=>{
const searchTerm=extractLastSearchTerm(url);
if(index==0)
{
    return result.concat(searchTerm);
}
const previousSearchTerm=result[result.length-1];
if(searchTerm===previousSearchTerm)
{
    return result;
}
else{
    return result.concat(searchTerm);
}
},[]).slice(-6).slice(0,-1);
const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState(
    'search',
    'React'
  );

  const [urls, setUrls] = React.useState([
    getUrl(searchTerm)
  ]);
  const handleSearch=searchTerm=>{
      const url=getUrl(searchTerm);
      setUrls(urls.concat(url));
  }

const handleLastSearches=searchTerm=>{
    setSearchTerm(searchTerm);

   handleSearch(searchTerm);
}

const lastSearches=getLastSearches(urls);

  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    { data: [], isLoading: false, isError: false }
  );

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: 'STORIES_FETCH_INIT' });

    try {
        const lastUrl=urls[urls.length-1];
      const result = await axios.get(lastUrl);

      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    }
  }, [urls]);

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = item => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  const handleSearchInput = event => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = event => {
    handleSearch(searchTerm);
    event.preventDefault();
  };

  return (
    <div className={styles.container}>
      <h1>My Hacker Stories</h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />


      <hr />

      {
          lastSearches.map((searchTerm,index)=>(
              <button key={searchTerm+index} type='button' onClick={()=>handleLastSearches(searchTerm)}>{searchTerm}</button>
          ))
      }
      {stories.isError && <p>Something went wrong ...</p>}

      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}) => (
  <form onSubmit={onSearchSubmit} className={styles.searchForm}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      isFocused
      onInputChange={onSearchInput}
    >
      <strong>Search:</strong>
    </InputWithLabel>

    <button type="submit" disabled={!searchTerm}>
      Submit
    </button>
  </form>
);

const InputWithLabel = ({
  id,
  value,
  type = 'text',
  onInputChange,
  isFocused,
  children,
}) => {
  const inputRef = React.useRef();

  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
      />
    </>
  );
};

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENT: list => sortBy(list, 'num_comments').reverse(),
  POINT: list => sortBy(list, 'points').reverse(),
};

const List = ({ list, onRemoveItem }) => {
  const [sort, setSort] = React.useState({
    sortKey: 'NONE',
    isReverse: false,
  });

  const handleSort = sortKey => {
    const isReverse = sort.sortKey === sortKey && !sort.isReverse;

    setSort({ sortKey, isReverse });
  };

  const sortFunction = SORTS[sort.sortKey];

  const sortedList = sort.isReverse
    ? sortFunction(list).reverse()
    : sortFunction(list);

  return (
    <div>
      <div style={{display:'flex'}}>
        <span style={{width:'40%'}}>TITLE</span>
        <span style={{width:'30%'}}>AUTHOR</span>
        <span style={{width:'10%'}}>COMENT</span>
        <span style={{width:'10%'}}>POINTS</span>
        <span style={{width:'10%'}}>ACTIONS</span>

      </div>
      <div style={{display:'flex'}}>
        <span  style={{width:'40%'}}>
          <button type="button" onClick={() => handleSort('TITLE')}>
            Title
          </button>
        </span>
        <span style={{width:'30%'}}> 
          <button type="button" onClick={() => handleSort('AUTHOR')}>
            Author
          </button>
        </span>
        <span style={{width:'10%'}}>
          <button type="button" onClick={() => handleSort('COMMENT')}>
            Comments
          </button>
        </span>
        <span style={{width:'10%'}}>
          <button type="button" onClick={() => handleSort('POINT')}>
            Points
          </button>
        </span >
        <span style={{width:'10%'}}><button type="button" onClick={() =>{}}>
            Actions
          </button></span>
      </div>

      {sortedList.map(item => (
        <Item
          key={item.objectID}
          item={item}
          onRemoveItem={onRemoveItem}
        />
      ))}
    </div>
  );
};

const Item = ({ item, onRemoveItem }) => (
  <div className={styles.item} style={{display:'flex'}}>
    <span style={{width:'40%'}}>
      <a href={item.url}>{item.title}</a>
    </span>
    <span style={{width:'30%'}}>{item.author}</span>
    <span style={{width:'10%'}}>{item.num_comments}</span>
    <span style={{width:'10%'}}>{item.points}</span>
    <span style={{width:'11%'}}>
      <button type="button" onClick={() => onRemoveItem(item)}>
        Dismiss
      </button>
    </span>
  </div>
);

export default App;




