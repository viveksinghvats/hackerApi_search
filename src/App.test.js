




import React from 'react';

import App,{storiesReducer,SearchForm,InputWithLabel,List,Item} from './App';

import {render,screen,fireEvent,act} from '@testing-library/react';
import { type } from 'os';

// import type {StoriesAction} from './App';
import { servicesVersion } from 'typescript';

import axios from 'axios';

// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });


jest.mock('axios');

// const mockedAxios = axios as jest.Mocked<typeof axios>;

const storyOne={
  title:'React',
  url:'https://react.js/org/',
  author:'Jordan Walke',
  num_comments:3,
  points:4,
  objectID:0
};

const storyTwo={
  title:'Redux',
  url:'https://redux.js/org/',
  author:'Dan Abramov, Andrew Clark',
  num_comments:2,
  points:5,
  objectID:1
};

describe('App',()=> {
  test('succeeds fetching data', async ()=>{
    const promise=Promise.resolve({
      data:{
        hits:stories,
      },
    }); 
  axios.get.mockImplementationOnce(()=> promise);
  render(<App/>);
  expect(screen.queryAllByText(/Loading/)).toBeInTheDocument();
  screen.debug();
  await act(()=>promise);
  // screen.debug();
  expect(screen.queryByText(/Loading/)).toBeNull();
  expect(screen.getByText('React')).toBeInTheDocument();
  expect(screen.getByText('Redux')).toBeInTheDocument();
  expect(screen.getAllByText('Dismiss').length).toBe(2);

  });
});






describe('something truthy and falsy',()=>{
  test('true to be true',()=>{
    expect(true).toBeTruthy();
  })
  test('false top be false',()=>{
    expect(true).toBe(true)
  })
});




const stories=[storyOne,storyTwo];
// interface StoriesRemoveAction{
//   type:'REMOVE_STORY';
// };
// type StoriesAction=|StoriesFetchInitAction|StoriesFetchSuccessAction|StoriesFetchFailureAction|StoriesRemoveAction;


describe('storiesReducer',()=>{
  test('removes a story from all stories',()=>{

    

    const action={ type: 'REMOVE_STORY', payload: storyOne }

   
    
const state={data:stories,isLoading:false,isError:false};
const newState=storiesReducer(state,action);
const expectedState={
  data:[storyTwo],
  isLoading:false,
  isError:false
};


expect(newState).toStrictEqual(expectedState);
  });
});

describe('Item',()=>{
  test('render all properties',()=>{
    render(<Item item={storyOne} OnRemoveItem={()=>{}} />);
    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
    expect(screen.getByText('React')).toHaveAttribute(
      'href','https://react.js/org/'
    );
    // screen.debug();
  });
  test('render a clickable dismiss button',()=>{
    render(<Item item={storyOne} OnRemoveItem={()=>{}}/>)
    expect(screen.getByRole('button')).toBeInTheDocument();
  });
  test('clicking the dismiss button call the callback handler',()=>{
    const handleRemoveItem=jest.fn();
    render(<Item item={storyOne} OnRemoveItem={handleRemoveItem}/>);
    fireEvent.click(screen.getByRole('button'));
    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
  });
});

describe('SearchForm',()=>{
  const searchFormProps={
    searchTerm:'React',
    onSearchSubmit:jest.fn(),
    onSearchInput:jest.fn()
  };

test('Render the input feilds with its value',()=>{
  render(<SearchForm {...searchFormProps}/>);
expect(screen.getByDisplayValue('React')).toBeInTheDocument();

  // screen.debug();
});

test('renders the  correct labels',()=>{
  render(<SearchForm {...searchFormProps}/>);
  expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
})

test('calls onSearchInput on input feild change',()=>{
  render(<SearchForm {...searchFormProps}/>);
  
  fireEvent.change(screen.getByDisplayValue('React'),{
    target:{
      value:'Redux'
    }
  });
  expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);

});

test('calls onSerachSubmit on button submit clicked',()=>{
  render(<SearchForm {...searchFormProps}/>);
  fireEvent.submit(screen.getByRole('button'));
   expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
});

test('renders snapshot',()=>{
  const {container} =render(<SearchForm {...searchFormProps}/>);
  expect(container.firstChild).toMatchSnapshot();
})

});



