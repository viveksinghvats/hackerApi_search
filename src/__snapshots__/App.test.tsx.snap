// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  class="searchForm"
>
  <label
    class="label"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="buttonbuttonLarge"
    type="submit"
  >
    Submit
  </button>
</form>
`;
